File: docs/Ref-Bulk/frontend/src/app/services/putaway.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';

export interface PutawayItem {
  lot_no: string;
  item_key: string;
  item_description?: string;
  location_key: string;
  bin_no?: string;
  qty_received: number;
  qty_on_hand: number;
  qty_available?: number; // New field from BME - calculated available quantity
  date_received: string;
  date_expiry: string;
  exp_date?: string; // New field from BME - formatted expiry date
  uom?: string; // New field from BME - Unit of Measure
  vendor_key: string;
  vendor_lot_no: string;
  document_no: string;
  lot_status: string;
  rec_user_id: string;
}

export interface PutawayRequest {
  lot_no: string;
  item_key: string;
  from_location: string;
  to_location: string;
  bin_no: string;
  qty_to_putaway: number;
  user_id: string;
}

export interface PutawayTransactionRequest {
  lot_no: string;
  item_key: string;
  from_location: string;
  to_location: string;
  from_bin?: string;
  to_bin: string;
  qty_to_putaway: number;
  user_id: string;
}

export interface PutawayTransactionResponse {
  success: boolean;
  message: string;
  transaction_id?: number;
  doc_no?: string;
  error_details?: string;
}

export interface PutawayConfirmation {
  success: boolean;
  message: string;
  transaction_id?: number;
}

export interface ScanRequest {
  barcode: string;
  scan_type: 'Item' | 'Location' | 'Lot';
}

export interface ScanResponse {
  valid: boolean;
  scan_type: 'Item' | 'Location' | 'Lot';
  data?: ScanData;
  message: string;
}

export interface ScanData {
  Item?: {
    item_key: string;
    description: string;
    unit: string;
    uom?: string; // Unit of Measure from BME
  };
  Location?: {
    location_key: string;
    description: string;
    location_type: string;
    available_bins?: string[]; // Available bins for BME search functionality
  };
  Lot?: {
    lot_no: string;
    item_key: string;
    qty_on_hand: number;
    qty_available?: number; // Available quantity from BME
    exp_date?: string; // Expiry date from BME
    uom?: string; // Unit of Measure from BME
  };
}

export interface PutawayHistory {
  transaction_id: number;
  lot_no: string;
  item_key: string;
  from_location: string;
  to_location: string;
  bin_no: string;
  qty_moved: number;
  transaction_date: string;
  user_id: string;
}

export interface TransactionAuditEntry {
  transaction_id: number;
  record_type: string;
  document_no: string;
  description: string;
  user_id: string;
  timestamp: string;
  details?: Record<string, unknown>;
}

export interface ApiResponse<T> {
  success: boolean;
  data?: T;
  message: string;
}

export interface PaginatedLotSearchResponse {
  items: LotSearchItem[];
  total: number;
  page: number;
  pages: number;
  limit: number;
}

@Injectable({
  providedIn: 'root'
})
export class PutawayService {
  private readonly baseUrl = environment.apiUrl || 'http://localhost:4400/api';

  constructor(private http: HttpClient) {}

  /**
   * Get all pending putaway items
   */
  getPendingPutawayItems(params?: { limit?: number; status?: string }): Observable<ApiResponse<PutawayItem[]>> {
    let httpParams = new HttpParams();
    if (params?.limit) {
      httpParams = httpParams.set('limit', params.limit.toString());
    }
    if (params?.status) {
      httpParams = httpParams.set('status', params.status);
    }

    return this.http.get<ApiResponse<PutawayItem[]>>(`${this.baseUrl}/putaway/pending`, {
      params: httpParams
    });
  }

  /**
   * Scan a barcode (item, location, or lot)
   */
  scanBarcode(request: ScanRequest): Observable<ApiResponse<ScanResponse>> {
    return this.http.post<ApiResponse<ScanResponse>>(`${this.baseUrl}/putaway/scan`, request);
  }

  /**
   * Confirm putaway operation (legacy)
   */
  confirmPutaway(request: PutawayRequest): Observable<ApiResponse<PutawayConfirmation>> {
    return this.http.post<ApiResponse<PutawayConfirmation>>(`${this.baseUrl}/putaway/confirm`, request);
  }

  /**
   * Confirm putaway transaction (enhanced with audit trail)
   */
  confirmPutawayTransaction(request: PutawayTransactionRequest): Observable<ApiResponse<PutawayTransactionResponse>> {
    return this.http.post<ApiResponse<PutawayTransactionResponse>>(`${this.baseUrl}/putaway/transaction`, request);
  }

  /**
   * Get putaway history
   */
  getPutawayHistory(limit?: number): Observable<ApiResponse<PutawayHistory[]>> {
    let httpParams = new HttpParams();
    if (limit) {
      httpParams = httpParams.set('limit', limit.toString());
    }

    return this.http.get<ApiResponse<PutawayHistory[]>>(`${this.baseUrl}/putaway/history`, {
      params: httpParams
    });
  }

  /**
   * Get specific putaway item details
   */
  getPutawayItemDetails(lotNo: string, itemKey: string): Observable<ApiResponse<PutawayItem>> {
    return this.http.get<ApiResponse<PutawayItem>>(`${this.baseUrl}/putaway/item/${lotNo}/${itemKey}`);
  }

  /**
   * Get transaction audit trail
   */
  getTransactionAudit(transactionId: number): Observable<ApiResponse<TransactionAuditEntry[]>> {
    return this.http.get<ApiResponse<TransactionAuditEntry[]>>(`${this.baseUrl}/putaway/audit/${transactionId}`);
  }

  /**
   * Search for lot numbers (BME functionality)
   */
  searchLotNumbers(query: string, limit: number = 10): Observable<ApiResponse<{ lot_no: string; item_key: string; item_description?: string }[]>> {
    const params = new HttpParams()
      .set('query', query)
      .set('limit', limit.toString());
    
    return this.http.get<ApiResponse<{ lot_no: string; item_key: string; item_description?: string }[]>>(`${this.baseUrl}/putaway/search/lots`, { params });
  }

  /**
   * Search for available bin numbers (BME functionality)
   */
  searchBinNumbers(locationKey: string, query: string, limit: number = 10): Observable<ApiResponse<{ bin_no: string; description?: string; status?: string }[]>> {
    const params = new HttpParams()
      .set('location', locationKey)
      .set('query', query)
      .set('limit', limit.toString());
    
    return this.http.get<ApiResponse<{ bin_no: string; description?: string; status?: string }[]>>(`${this.baseUrl}/putaway/search/bins`, { params });
  }

  /**
   * Get available bins for a location (BME functionality)
   */
  getAvailableBins(locationKey: string): Observable<ApiResponse<{ bin_no: string; description?: string; capacity?: number; available_space?: number }[]>> {
    const params = new HttpParams().set('location', locationKey);
    
    return this.http.get<ApiResponse<{ bin_no: string; description?: string; capacity?: number; available_space?: number }[]>>(`${this.baseUrl}/putaway/bins/available`, { params });
  }

  // ========================================================================================
  // ENHANCED PUTAWAY API - Replicates Official App Functionality (BT-25268027 Pattern)
  // ========================================================================================

  /**
   * Search for lot details using clean API
   * Replicates official app lot search functionality
   */
  searchLot(lotNo: string): Observable<LotSearchResponse> {
    return this.http.get<LotSearchResponse>(`${this.baseUrl}/putaway/lot/${lotNo}`);
  }

  /**
   * Validate destination bin using clean API
   * Replicates official app bin validation functionality
   */
  validateBin(location: string, binNo: string): Observable<BinValidationResponse> {
    return this.http.get<BinValidationResponse>(`${this.baseUrl}/putaway/bin/${location}/${binNo}`);
  }

  /**
   * Execute bin transfer using clean API
   * Replicates official app BT-25268027 transaction pattern
   */
  executeBinTransfer(request: BinTransferRequest): Observable<TransactionResponse> {
    return this.http.post<TransactionResponse>(`${this.baseUrl}/putaway/transfer`, request);
  }

  /**
   * Search multiple lots with optional query filter
   * Replicates BME lot selection dialog functionality
   */
  searchMultipleLots(query?: string, limit: number = 20): Observable<LotSearchItem[]> {
    let params = new HttpParams();
    if (query) {
      params = params.set('query', query);
    }
    params = params.set('limit', limit.toString());
    
    return this.http.get<LotSearchItem[]>(`${this.baseUrl}/putaway/lots/search`, { params });
  }

  /**
   * Search multiple lots with pagination support
   * Enhanced version with server-side pagination
   */
  searchMultipleLotsWithPagination(query?: string, page: number = 1, limit: number = 20): Observable<PaginatedLotSearchResponse> {
    let params = new HttpParams();
    if (query) {
      params = params.set('query', query);
    }
    params = params.set('page', page.toString());
    params = params.set('limit', limit.toString());
    
    return this.http.get<PaginatedLotSearchResponse>(`${this.baseUrl}/putaway/lots/search`, { params });
  }

  /**
   * Search for bins with pagination support
   * Enhanced version with server-side pagination
   */
  searchBinsWithPagination(query?: string, page: number = 1, limit: number = 20): Observable<PaginatedBinSearchResponse> {
    let params = new HttpParams();
    if (query) {
      params = params.set('query', query);
    }
    params = params.set('page', page.toString());
    params = params.set('limit', limit.toString());
    
    return this.http.get<PaginatedBinSearchResponse>(`${this.baseUrl}/putaway/bins/search`, { params });
  }

  /**
   * Get putaway service health status
   */
  getPutawayHealth(): Observable<HealthResponse> {
    return this.http.get<HealthResponse>(`${this.baseUrl}/putaway/health`);
  }
}

// ========================================================================================
// CLEAN PUTAWAY API INTERFACES - Official App Compatibility
// ========================================================================================

export interface LotSearchResponse {
  lot_no: string;
  item_key: string;
  location: string;
  current_bin: string;
  qty_on_hand: number;
  qty_available: number;
  expiry_date?: string;
  item_description: string;
  uom: string;
  lot_status: string;
}

export interface BinValidationResponse {
  bin_no: string;
  location: string;
  is_valid: boolean;
  message: string;
}

export interface BinTransferRequest {
  lot_no: string;
  item_key: string;
  location: string;
  bin_from: string;
  bin_to: string;
  transfer_qty: number;
  user_id: string;
  remarks?: string;
  referenced?: string;
}

export interface TransactionResponse {
  success: boolean;
  document_no: string;
  message: string;
  timestamp: string;
}

export interface HealthResponse {
  status: string;
  service: string;
  timestamp: string;
  version: string;
}

export interface LotSearchItem {
  lot_no: string;
  item_key: string;
  item_description: string;
  location: string;
  current_bin: string;
  qty_on_hand: number;
  qty_available: number;
  expiry_date?: string;
  uom: string;
  lot_status: string;
}

export interface BinSearchItem {
  bin_no: string;
  location: string;
  description: string;
  aisle: string;
  row: string;
  rack: string;
}

export interface PaginatedBinSearchResponse {
  items: BinSearchItem[];
  total: number;
  page: number;
  pages: number;
  limit: number;
}

File: docs/Ref-Bulk/backend/src/services/putaway_service.rs
use crate::utils::bangkok_now_rfc3339;
use crate::database::{Database, putaway_db::PutawayDatabase};
use crate::models::putaway_models::{
    LotSearchResult, BinValidationResult, BinTransferRequest, 
    TransferResult, PutawayHealthResponse, LotSearchItem, BinSearchItem, PutawayError
};

pub struct PutawayService {
    db: PutawayDatabase,
}

impl PutawayService {
    pub fn new(database: Database) -> Self {
        Self {
            db: PutawayDatabase::new(database),
        }
    }

    /// Search for lot details by lot number
    pub async fn search_lot(&self, lot_no: &str) -> Result<LotSearchResult, PutawayError> {
        // Validate input
        if lot_no.trim().is_empty() {
            return Err(PutawayError::ValidationError("Lot number cannot be empty".to_string()));
        }

        // Search in database
        match self.db.find_lot_by_number(lot_no).await? {
            Some((lot_record, item_record)) => {
                // Calculate available quantity (QtyOnHand - QtyCommitSales)
                let qty_available = lot_record.qty_on_hand - lot_record.qty_commit_sales;

                // Format expiry date
                let expiry_date = lot_record.date_expiry.format("%Y-%m-%d").to_string();

                Ok(LotSearchResult {
                    lot_no: lot_record.lot_no,
                    item_key: lot_record.item_key,
                    location: lot_record.location_key,
                    current_bin: lot_record.bin_no,
                    qty_on_hand: lot_record.qty_on_hand,
                    qty_available,
                    expiry_date: Some(expiry_date),
                    item_description: item_record.desc1,
                    uom: item_record.stock_uom_code,
                    lot_status: lot_record.lot_status,
                })
            }
            None => Err(PutawayError::LotNotFound { lot_no: lot_no.to_string() }),
        }
    }

    /// Validate destination bin
    pub async fn validate_bin(&self, location: &str, bin_no: &str) -> Result<BinValidationResult, PutawayError> {
        // Validate input
        if location.trim().is_empty() || bin_no.trim().is_empty() {
            return Ok(BinValidationResult {
                bin_no: bin_no.to_string(),
                location: location.to_string(),
                is_valid: false,
                message: "Location and bin number cannot be empty".to_string(),
            });
        }

        // Check if bin is valid
        match self.db.validate_bin_location(location, bin_no).await {
            Ok(is_valid) => {
                let message = if is_valid { "Bin is valid and available".to_string() } else { format!("Bin '{bin_no}' not found in location '{location}'") };

                Ok(BinValidationResult {
                    bin_no: bin_no.to_string(),
                    location: location.to_string(),
                    is_valid,
                    message,
                })
            }
            Err(e) => Err(e),
        }
    }

    /// Execute bin transfer using validated/corrected quantity
    pub async fn execute_transfer(&self, request: BinTransferRequest) -> Result<TransferResult, PutawayError> {
        // Validate request
        self.validate_transfer_request(&request)?;

        // Validate in database and get corrected transfer quantity for full transfers
        let (actual_transfer_qty, is_full_transfer) = self.db.validate_transfer_request(
            &request.lot_no,
            &request.item_key,
            &request.location,
            &request.bin_from,
            &request.bin_to,
            request.transfer_qty,
        ).await?;

        // Execute transfer with the corrected quantity (exact available qty for full transfers)
        match self.db.execute_bin_transfer_transaction(
            &request.lot_no,
            &request.item_key,
            &request.location,
            &request.bin_from,
            &request.bin_to,
            actual_transfer_qty,
            &request.user_id,
            request.remarks.as_deref().unwrap_or(""),
            request.referenced.as_deref().unwrap_or(""),
        ).await {
            Ok(document_no) => {
                Ok(TransferResult {
                    success: true,
                    document_no,
                    message: if is_full_transfer {
                        format!(
                            "Successfully transferred {} units (FULL TRANSFER) of lot {} from {} to {} - Source bin cleared",
                            actual_transfer_qty, request.lot_no, request.bin_from, request.bin_to
                        )
                    } else {
                        format!(
                            "Successfully transferred {} units of lot {} from {} to {}",
                            actual_transfer_qty, request.lot_no, request.bin_from, request.bin_to
                        )
                    },
                    timestamp: bangkok_now_rfc3339(),
                })
            }
            Err(e) => {
                Ok(TransferResult {
                    success: false,
                    document_no: String::new(),
                    message: format!("Transfer failed: {e}"),
                    timestamp: bangkok_now_rfc3339(),
                })
            }
        }
    }


    /// Search for lots with pagination
    pub async fn search_lots_paginated(&self, query: Option<&str>, page: i32, limit: i32) -> Result<(Vec<LotSearchItem>, i32), PutawayError> {
        // Validate inputs
        let safe_page = if page < 1 { 1 } else { page };
        let safe_limit = if limit > 100 { 100 } else if limit < 1 { 20 } else { limit };
        
        // Search in database with pagination
        self.db.search_lots_paginated(query, safe_page, safe_limit).await
    }

    /// Get service health status
    pub async fn get_health(&self) -> PutawayHealthResponse {
        PutawayHealthResponse {
            status: "healthy".to_string(),
            service: "putaway".to_string(),
            timestamp: bangkok_now_rfc3339(),
            version: "1.0.0".to_string(),
        }
    }

    /// Validate transfer request fields
    fn validate_transfer_request(&self, request: &BinTransferRequest) -> Result<(), PutawayError> {
        if request.lot_no.trim().is_empty() {
            return Err(PutawayError::ValidationError("Lot number is required".to_string()));
        }

        if request.item_key.trim().is_empty() {
            return Err(PutawayError::ValidationError("Item key is required".to_string()));
        }

        if request.location.trim().is_empty() {
            return Err(PutawayError::ValidationError("Location is required".to_string()));
        }

        if request.bin_from.trim().is_empty() {
            return Err(PutawayError::ValidationError("Source bin is required".to_string()));
        }

        if request.bin_to.trim().is_empty() {
            return Err(PutawayError::ValidationError("Destination bin is required".to_string()));
        }

        if request.bin_from == request.bin_to {
            return Err(PutawayError::ValidationError("Source and destination bins cannot be the same".to_string()));
        }

        if request.transfer_qty <= 0.0 {
            return Err(PutawayError::ValidationError("Transfer quantity must be greater than zero".to_string()));
        }

        if request.user_id.trim().is_empty() {
            return Err(PutawayError::ValidationError("User ID is required".to_string()));
        }

        Ok(())
    }

    /// Search for bins with pagination
    pub async fn search_bins_paginated(&self, query: Option<&str>, page: i32, limit: i32) -> Result<(Vec<BinSearchItem>, i32), PutawayError> {
        // Validate inputs
        let safe_page = if page < 1 { 1 } else { page };
        let safe_limit = if limit > 100 { 100 } else if limit < 1 { 20 } else { limit };
        
        // Search in database with pagination
        self.db.search_bins_paginated(query, safe_page, safe_limit).await
    }
}

